### ISSUE TITLE
-h should be exempt from "shorthand has been deprecated" warning

### ISSUE THREAD

**Issue by jrose-signal** [docker/cli#3249]
I was trying to learn `docker build`'s options today, so I put `docker build -h` into my shell and got the following message (and then the help output).

```
Flag shorthand -h has been deprecated, please use --help
```

For most flags, this seems reasonable, but I typed `-h` because I don't *know* what Docker's argument conventions are. The help output even lists a number of short flags that are presumably still valid (`-f file`, `-o destination`, `-q`, `-t tag`). `-h` should be one of them, possibly even with an extra special case for `-help`.

**Comment by thaJeztah** [comment]
It's only the `-h` shorthand flag that's deprecated, not all shorthand flags; https://docs.docker.com/engine/deprecated/#-h-shorthand-for---help

**Comment by jrose-signal** [comment]
Oh. That's even worse.

**Comment by thaJeztah** [comment]
Can you describe the problem you ran into? Other than printing the message at the top, it should still print the same information, e.g. 

```bash
docker build -h
Flag shorthand -h has been deprecated, please use --help

Usage:  docker build [OPTIONS] PATH | URL | -

Build an image from a Dockerfile

Options:
      --add-host list           Add a custom host-to-IP mapping (host:ip)
      --build-arg list          Set build-time variables
      --cache-from strings      Images to consider as cache sources
      --disable-content-trust   Skip image verification (default true)
  -f, --file string             Name of the Dockerfile (Default is 'PATH/Dockerfile')
      --iidfile string          Write the image ID to the file
      --isolation string        Container isolation technology
      --label list              Set metadata for an image
      --network string          Set the networking mode for the RUN instructions during build (default "default")
      --no-cache                Do not use cache when building the image
  -o, --output stringArray      Output destination (format: type=local,dest=path)
      --platform string         Set platform if server is multi-platform capable
      --progress string         Set type of progress output (auto, plain, tty). Use plain to show container output (default "auto")
      --pull                    Always attempt to pull a newer version of the image
  -q, --quiet                   Suppress the build output and print image ID on success
      --secret stringArray      Secret file to expose to the build (only if BuildKit enabled): id=mysecret,src=/local/secret
      --ssh stringArray         SSH agent socket or keys to expose to the build (only if BuildKit enabled) (format: default|<id>[=<socket>|<key>[,<key>]])
  -t, --tag list                Name and optionally a tag in the 'name:tag' format
      --target string           Set the target build stage to build.
```

**Comment by jrose-signal** [comment]
Yes, it did (sorry for not being clearer about that!). The problem is that someone new to Docker is being told they're using the CLI incorrectly *while they're trying to learn how to use it.* It's an opportunity the program had to be accomodating of new users and instead it scolds them in their first command with a deprecation they most likely had no way of knowing about. I'd suggest going the other way, that even in the subcommands that already use `-h hostname` a trailing `-h` should be seen as an alias for `--help`.

**Comment by thaJeztah** [comment]
Changing `docker run -h <hostname> ... ` to become `--help` will unfortunately be a breaking change, which is not something we can do. While `-h` is common, not all tools support this (`--help` is often used as well), and at least not part of the [POSIX guidelines](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html).

**Comment by jrose-signal** [comment]
I just meant in cases of `docker run -h` or `docker run --foo --bar baz.image -h`, where the command would already not parse successfully. Those cases are very unlikely to be someone trying to use a hostname. The POSIX guidelines don't mention `--help` either, so I wouldn't really take that as evidence one way or another.

**Comment by Nadock** [comment]
Just saw this warning for the first time today — I'm not sure what the expected gain is for removing `-h`.

First of all, it's a really annoying hit to the UX of the CLI. I know for me and a lot of other developers, the first thing we do when we're not sure about a CLI command is to run `${cmd} -h`. Removing this and forcing someone to type out the long version just kinda sucks.

But beyond that, what will be the behaviour of `docker -h` after this is removed? If it prints the CLI help, then you haven't actually removed anything and if the CLI _tells_ you to run with `--help` why not just print the help and save the user a step? 

[`-h`/`--help` is good CLI design.](https://clig.dev/#help)

**Comment by fenugrec** [comment]
Another reason to keep  `-h`  - you use that a lot when refreshing on less usual commands, args, etc. This means manually typing that extra `-help` possibly many times in a row - extremely annoying and tedious...
